File: 0648-replace-words/0648-replace-words.cpp
class Solution {
public:
    string replaceWords(vector<string>& dictionary, string sentence) {
    unordered_set<string> dict(dictionary.begin(), dictionary.end());
    stringstream ss(sentence);
    string word;
    string result;

    while (ss >> word) {
        string root;
        for (int i = 1; i <= word.length(); ++i) {
            root = word.substr(0, i);
            if (dict.find(root) != dict.end()) {
                break;
            }
        }
        if (!result.empty()) {
            result += " ";
        }
        result += root;
    }

    return result;
        
    }
};

File: 0648-replace-words/0648-replace-words.ts
function replaceWords(dictionary: string[], sentence: string): string {
    const sent = sentence.split(' ');
    const dict = new Set(dictionary);
    let lastIndex = sent.length - 1;
    let ans = "";
    for(let word of sent){
        let i = 0;
        while(i < word.length){
            let root = word.slice(0,i+1);
            if(dict.has(root)){
                ans += ` ${root}`;
                break;
            }
            else if(!dict.has(root) && i==word.length - 1){
                ans += ` ${word}`;
            }
            i++;
        }        
    }
    return ans.trim();
};